### PR TITLE
Introduce (procarg0) node for tCOMMA in "|a,|".

### LINKED ISSUE #312: Trailing commas for block arguments should be parsed as mlhs
Ruby lets you use a trailing comma in a block argument for destructuring

```ruby
people = { 'John' => 22, 'Jim' => 23 }

people.each do |name,|
  puts "Hello, #{name}!"
end

# >> Hello, John!
# >> Hello, Jim!
```

Removing the comma results in different behavior:

```ruby
people = { 'John' => 22, 'Jim' => 23 }

people.each do |name|
  puts "Hello, #{name}!"
end

# >> Hello, ["John", 22]!
# >> Hello, ["Jim", 23]!
```

Parser produces the same AST for both cases:

```sh
$ ruby-parse --version
ruby-parse based on parser version 2.3.1.2

$ ruby-parse -e 'foo.each { |bar,| bar }'
(block
  (send
    (send nil :foo) :each)
  (args
    (arg :bar))
  (lvar :bar))

$ ruby-parse -e 'foo.each { |bar| bar }'
(block
  (send
    (send nil :foo) :each)
  (args
    (arg :bar))
  (lvar :bar))
```

## Expected behavior

Parser correctly handles destructuring like this for assignments:

```sh
$ ruby-parse -e 'foo = bar'
(lvasgn :foo
  (send nil :bar))
$ ruby-parse -e 'foo, = bar'
(masgn
  (mlhs
    (lvasgn :foo))
  (send nil :bar))
```

and parser also supports other forms of destructuring for block arguments:

```sh
$ ruby-parse -e 'foo.each { |(bar)| bar }'
(block
  (send
    (send nil :foo) :each)
  (args
    (mlhs
      (arg :bar)))
  (lvar :bar))
```

so it seems like the appropriate fix would be a parse that wraps the block arguments in `(mlhs ...)` like so:

```sh
$ ruby-parse -e 'foo.each { |bar,| bar }'
(block
  (send
    (send nil :foo) :each)
  (args
    (mlhs
      (arg :bar)))
  (lvar :bar))
```

Thanks for your great work on parser! I would be happy to look into this bug later today when I have some extra time.

### PR DESCRIPTION
**Pull request by iliabylich**
Closes #312.

### CODE REVIEW

**Comment by whitequark** [comment]
I prefer the name `procarg0`. Also, the AST should look like `s(:args, s(:procarg0, :a))` for the `m { |a,| }` case since this is logically equivalent to a sort of mlhs.

Also, this is a somewhat exceptional case in that the comma is not just a separator but it also has semantic meaning. So, we need to put the source location of the comma into the source map. I think `Source::Map::Operator` would do here.

**Comment by whitequark** [comment]
Oh, and you should update `doc/AST_FORMAT.md`.

**Comment by iliabylich** [comment]
@whitequark Should this `procarg0` node be an operator? And if so, should I document it in the `Arguments` section of `doc/AST_FORMAT.md`?

**Comment by whitequark** [comment]
> Should this procarg0 node be an operator?

It is an argument node, like `s(:kwarg)`. It should be documented in that section. The reason it uses `Source::Map::Operator` is distinct; the correspondence from AST nodes to map classes is ad-hoc and doesn't carry much semantic meaning (neither it is even stable, only the valid accessors are).

**Comment by iliabylich** [comment]
@whitequark Done I guess.

File: doc/AST_FORMAT.md [hunk 1]
```diff
@@ -997,6 +997,17 @@ Format:
...
 ~~~
 
+### "Excessed comma"
+
+Representation of the trailing comma after block arguments.
+
+~~~
+(procarg0)
+"|a,|"
+ ~~~~ expression
```
Comment (whitequark) [inline doc/AST_FORMAT.md]: Should be:
```
|a,|
 ~~ expression
  ~ operator
```

File: lib/parser/builders/default.rb [hunk 2]
```diff
@@ -24,9 +24,28 @@ class << self
...
+      # `s(:args, s(:arg, :a))`
+      #
+      # If set to true, arguments of `m { |a,| }` are emitted as
+      # `s(:args, s(:arg, :a), s(:procarg0))
+      #
+      # @return [Boolean]
+      attr_accessor :emit_excessed_comma
+    end
+
+    @emit_excessed_comma = false
```
Comment (whitequark) [inline lib/parser/builders/default.rb]: `@emit_procarg0` and elsewhere.

File: test/test_parser.rb [hunk 3]
```diff
@@ -3388,6 +3383,26 @@ def test_send_lambda_legacy
     Parser::Builders::Default.emit_lambda = true
   end
 
+  def test_excessed_comma
+    assert_parses(
+      s(:block, s(:send, nil, :m),
+        s(:args, s(:arg, :a), s(:procarg0)), nil),
+      %q{m { |a, | }},
+      %q{    ~~~~~ expression (args)
```
Comment (whitequark) [inline test/test_parser.rb]: needs a test for `expression (args.procarg0)`.

File: test/test_parser.rb [hunk 4]
```diff
@@ -3388,6 +3383,26 @@ def test_send_lambda_legacy
...
+        |      ~ operator (args.procarg0)
+      }
+    )
+  end
+
+  def test_excessed_comma_legacy
+    Parser::Builders::Default.emit_excessed_comma = false
+    assert_parses_blockargs(
+      s(:args, s(:arg, :a)),
+      %q{|a,|}
```
Comment (whitequark) [inline test/test_parser.rb]: also needs a location test

File: doc/AST_FORMAT.md (hunk 1 above)
Comment (iliabylich) [inline doc/AST_FORMAT.md]: But I'm getting this test failing locally:
``` ruby
assert_parses(
  s(:block, s(:send, nil, :m),
    s(:args, s(:arg, :a)), nil),
  %q{m { |a| }},
  %q{     ~ expression (args)
  }
)
```
While this one passes:
``` ruby
assert_parses(
  s(:block, s(:send, nil, :m),
    s(:args, s(:arg, :a)), nil),
  %q{m { |a| }},
  %q{    ~~~ expression (args)
  }
)
```

File: test/test_parser.rb (hunk 4 above)
Comment (iliabylich) [inline test/test_parser.rb]: What should be tested here? There's no `procarg0` node.

File: doc/AST_FORMAT.md (hunk 1 above)
Comment (whitequark) [inline doc/AST_FORMAT.md]: No no, the expression for `(args)` encompasses `|a|` whereas the expression for `(args.arg)` encompasses `a`.

**Comment by iliabylich** [comment]
Finally got it, fixed :smile:

**Comment by whitequark** [comment]
As described in https://github.com/whitequark/parser/issues/312#issuecomment-252407424, the fix in this PR is essentially inverted from the real semantics. Can you please make it so that:

  * `proc{|a|}` results in `(args (procarg0 :a))`;
  * everything else results in exactly the same code as before, i.e. `proc{|a,|}` becomes `(args (arg :a))`.

There should probably be some sort of flag passed to the builder. I suggest first altering `ruby24.y`, and once that passes review you can duplicate it elsewhere.

**Comment by iliabylich** [comment]
@whitequark Is it possible to handle it on the grammar level? I mean, `lambda` and `proc` are just Kernel methods, they can be aliased to each other or completely removed. For me it sounds like a method-level grammar, the only block that has a new argument type is a block passed to a `proc` method and has **exactly** one argument without trailing comma.

**Comment by whitequark** [comment]
> I mean, lambda and proc are just Kernel methods, they can be aliased to each other or completely removed.

This is wrong. Lambda/proc distinction, although visible through the Kernel methods, is fundamental to the language. There are not just two methods, but two distinct types of closures in Ruby;  `-> {}` is an alias for `lambda` (you used to be able to override this by replacing `Kernel#lambda`, and this is why we didn't originally have a `(lambda)` node, just a `(block (send (const :Kernel) :lambda) etc)`, but not anymore), whereas `do f(&b) b end; f {}` is an alias for `proc`.

```
> ->{}.lambda?
=> true
> def f(&b) b end; f {}.lambda?
=> false
```

The thing is that the parser has exactly one syntactic construct, a block, but depending on which runtime context this block appears in, it can become a proc and a lambda. Moreover these two are not interconvertible into each other:
```
[9] pry(main)> proc(&->{}).lambda?
=> true
[11] pry(main)> lambda(&f{}).lambda?
=> false
```

Anyway, the bottom line is that when we parse the block--when we are *about to pass* a block to a function--we have no idea if the arguments should be proc-like or lambda-like, and so this information has to be preserved in the AST.

**Comment by whitequark** [comment]
> This is wrong.

Scratch this, I may have just misunderstood you.

**Comment by whitequark** [comment]
Let me try to rephrase the above the other way, and maybe more concisely: it is not possible to handle the distinction *purely* on the AST/grammar level, however a marker on AST/grammar level is *necessary* for the runtime support code to decide on how the arguments should be parsed, in case the block is passed to `proc`.

**Comment by iliabylich** [comment]
@whitequark Consider the following example:
``` ruby
module Kernel
  alias proc2 proc
  alias lambda2 lambda
end

[
  lambda { |a,| [a] },
  lambda2 { |a,| [a] },

  proc { |a,| [a] },
  proc2 { |a,| [a] }
].each do |pr|
  p pr.call([1])
end
```

I'm trying to say that we can't simply stick to a method called `proc`. `proc` is just a method name that has some behavior and this method can potentially have absolutely any name.

**Comment by whitequark** [comment]
> I'm trying to say that we can't simply stick to a method called proc. proc is just a method name that has some behavior and this method can potentially have absolutely any name.

Oh! I agree of course, I wrote my earlier comment misleadingly. In my earlier comment...

> proc{|a|} results in (args (procarg0 :a));
> everything else results in exactly the same code as before, i.e. proc{|a,|} becomes (args (arg :a)).

by `proc` I mean "any expression", not the `proc` function specifically.

**Comment by iliabylich** [comment]
Alright, updated only 2.4, other versions are temporary excluded from the build/test suite.

@whitequark Could you review it please? The code is ugly but at least is seems to be correct now.

**Comment by whitequark** [comment]
I think you can get away with making just one change, in the `block_param: f_arg opt_block_args_tail` rule. That is, when val[1] is empty and val[0] has one element, branch to `@builder.procarg0` else do like before.

**Comment by iliabylich** [comment]
@whitequark Great idea, done. Any other comments before applying same rules on other versions?
By the way, I guess `AST::Processor` also requires one more alias for a new argument type - is it true?

**Comment by whitequark** [comment]
> Any other comments before applying same rules on other versions?

Looks good to me.

> By the way, I guess AST::Processor also requires one more alias for a new argument type - is it true?

Yep.

**Comment by iliabylich** [comment]
What should happen with `(arg_expr)` from 1.8? One more node?
As far as I understand it should be:

``` ruby
proc { |@a| }

# Before
(block
  (send nil :proc)
  (args
    (arg-expr
      (ivasgn :@a))) nil)

# After
(block
  (send nil :proc)
  (args
    (procarg0-expr
      (ivasgn :@a))) nil)
```

**Comment by whitequark** [comment]
Nothing.
```
[10] pry(main)> proc{|a,|[a]}.call(1)
=> [1]
[11] pry(main)> proc{|a,|[a]}.call([1])
=> [[1]]
[12] pry(main)> proc{|a|[a]}.call(1)                                                                                                           
=> [1]                                                                                                                                            
[13] pry(main)> proc{|a|[a]}.call([1])                                                                                                            
=> [[1]]                                                                                                                                          
[14] pry(main)> RUBY_VERSION                                                                                                                      
=> "1.8.7"                                                                                                                                        
```

**Comment by iliabylich** [comment]
@whitequark Was about to ask you for a final review, but then realized that this fix doesn't handle single optional argument:
``` ruby
2.3.0 :001 > proc { |a=1| a }.call(1)
 => 1 
2.3.0 :002 > proc { |a=1| a }.call([1])
 => [1] 
2.3.0 :003 > proc { |a| a }.call(1)
 => 1 
2.3.0 :004 > proc { |a| a }.call([1])
 => [1]
```

I believe it can't be solved without introducing 1 more node. `(optprocarg0`) sounds horrible.

How about closing this issue? I'm sure that it can be handled by the library that uses `parser` using source maps. [This](https://github.com/opal/opal/blob/master/lib/opal/nodes/iter.rb#L167-L172) is how opal handles it atm.

I just think that it's a bad idea to break compatibility and introduce some new nodes to handle such a small issue.

**Comment by whitequark** [comment]
> How about closing this issue? I'm sure that it can be handled by the library that uses parser using source maps. This is how opal handles it atm.

No. This is utter nonsense. Parser's AST is designed such that source maps are *never* required to interpret them semantically. `optprocarg0` looks just fine to me.

> I just think that it's a bad idea to break compatibility and introduce some new nodes to handle such a small issue.

We don't break compatibility because this is opt-in.

**Comment by iliabylich** [comment]
@whitequark Added `optprocarg0` node, could you check it please?

File: doc/AST_FORMAT.md [hunk 5]
```diff
@@ -864,6 +864,19 @@ Format:
  ~~~ name
 ~~~
 
+### Required single block argument
```
Comment (whitequark) [inline doc/AST_FORMAT.md]: Just put the version requirement in the title, like `### Required single block argument (1.9)`. This is the convention used elsewhere.

File: doc/AST_FORMAT.md [hunk 6]
```diff
@@ -864,6 +864,19 @@ Format:
...
+
+Only for parser > 1.8.
+
+Format:
+
+~~~
+(procarg0 :foo)
+|foo|
```
Comment (whitequark) [inline doc/AST_FORMAT.md]: This should be in quotes, like `"|foo|"`.

File: doc/AST_FORMAT.md [hunk 7]
```diff
@@ -876,6 +889,17 @@ Format:
  ~~~ name
 ~~~
 
+### Optional single block argument
```
Comment (whitequark) [inline doc/AST_FORMAT.md]: I think both of these should go just before "Expression arguments" and gain a small note about their difference from `|a,|`.

**Comment by iliabylich** [comment]
Hold on, MRI doesn't let you to pass trailing comma after optional block argument. Do we really need a `optprocarg0` node? There are no places where you need to identify a trailing comma after optional argument and make it acting like mlhs *because it's just impossible*.

**Comment by whitequark** [comment]
Then nope!

**Comment by iliabylich** [comment]
Updated PR

**Comment by iliabylich** [comment]
Thank you for your patience :smile:

**Pull request merged**